lib/Geo/GeoJson.cpp

```diff
@@ -39,616 +41,765 @@
-        return parseMultiPoint(vpack, region);
-      }
-      case Type::LINESTRING: {
-        auto line = std::make_unique<S2Polyline>();
-        Result res = parseLinestring(vpack, *line.get());
-        if (res.ok()) {
-          region.reset(std::move(line), ShapeContainer::Type::S2_POLYLINE);
-        }
-        return res;
-      }
-      case Type::MULTI_LINESTRING: {
-        std::vector<S2Polyline> polylines;
-        Result res = parseMultiLinestring(vpack, polylines);
-        if (res.ok()) {
-          region.reset(new S2MultiPolyline(std::move(polylines)),
-                       ShapeContainer::Type::S2_MULTIPOLYLINE);
-        }
-        return res;
-      }
-      case Type::POLYGON: {
-        return parsePolygon(vpack, region, legacy);
-      }
-      case Type::MULTI_POLYGON: {
-        return parseMultiPolygon(vpack, region, legacy);
-      }
-      case Type::GEOMETRY_COLLECTION: {
-        return Result(TRI_ERROR_NOT_IMPLEMENTED,
-                      "GeoJSON type is not supported");
-      }
-      case Type::UNKNOWN: {
-        // will return TRI_ERROR_BAD_PARAMETER
-        break;
-      }
-    }
-  }
-
-  return Result(TRI_ERROR_BAD_PARAMETER, "Invalid GeoJSON Geometry Object.");
+  return {};
 }
 
-/// @brief create s2 latlng
-Result parsePoint(VPackSlice const& vpack, S2LatLng& latLng) {
-  if (Type::POINT != type(vpack)) {
-    return {TRI_ERROR_BAD_PARAMETER, "require type: 'Point'"};
+template<bool Validation>
+Result parsePointsImpl(velocypack::Slice vpack,
+                       std::vector<S2Point>& vertices) {
+  auto r = parseImpl<Validation, true>(vpack, vertices);
+  if (Validation && ADB_UNLIKELY(!r.ok())) {
+    return r;
   }
-  TRI_ASSERT(Type::POINT == type(vpack));
-
-  VPackSlice coordinates = vpack.get(Fields::kCoordinates);
-  if (coordinates.isArray() && coordinates.length() == 2 &&
-      coordinates.at(0).isNumber() && coordinates.at(1).isNumber()) {
-    latLng = S2LatLng::FromDegrees(coordinates.at(1).getNumber<double>(),
-                                   coordinates.at(0).getNumber<double>())
-                 .Normalized();
-    return TRI_ERROR_NO_ERROR;
+  if (Validation && ADB_UNLIKELY(vertices.empty())) {
+    return {TRI_ERROR_BAD_PARAMETER,
+            "Invalid MultiPoint, it must contains at least one point."};
   }
-  return TRI_ERROR_BAD_PARAMETER;
+  return {};
 }
 
-/// https://tools.ietf.org/html/rfc7946#section-3.1.3
-Result parseMultiPoint(VPackSlice const& vpack, ShapeContainer& region) {
-  if (!vpack.isObject()) {
-    return Result(TRI_ERROR_BAD_PARAMETER, "Invalid GeoJSON Geometry Object.");
+template<bool Validation>
+Result parseLineImpl(velocypack::Slice vpack, std::vector<S2Point>& vertices) {
+  auto r = parseImpl<Validation, true>(vpack, vertices);
+  if (Validation && ADB_UNLIKELY(!r.ok())) {
+    return r;
   }
-
-  if (Type::MULTI_POINT != type(vpack)) {
-    return TRI_ERROR_BAD_PARAMETER;
+  removeAdjacentDuplicates(vertices);
```

Comment:
Do we need this call in case of Validation=false? I guess in this case we expect that someone already called this while parsing with Validation=true before. Or we store "non-normalized" version? Then it is ok to leave it here.

Reply:
Yes :( As I wrote with Validation false we don't need Validation but still need transforming from valid vpack GeoJson to valid s2 input